tools/hdf_dump_translation_dataset.py

```diff
@@ -0,0 +1,280 @@
+    target_lines = islice(self.target_file_handle, self.line_buffer_size)
+
+    source_chunk = []
+    target_chunk = []
+    for source_line, target_line in zip_longest(source_lines, target_lines):
+      assert source_line is not None and target_line is not None, "Number of source and target lines differ."
+
+      source_indices = self._line_to_indices(source_line, "source")
+      target_indices = self._line_to_indices(target_line, "target")
+
+      source_chunk.append(source_indices)
+      target_chunk.append(target_indices)
+
+    return source_chunk, target_chunk
+
+  def _write_lengths(self, source_lengths, target_lengths):
+    lengths = numpy.array([source_lengths, target_lengths], dtype="int32").transpose()
+
+    offset = self._number_of_processed_lines
+    assert len(lengths) + offset <= self.number_of_lines, "More lines in the corpus files than specified."
+
+    self.hdf_file["seqLengths"][offset:offset + len(lengths), :] = lengths
+
+  def _write_tags(self, chunk_size):
+    offset = self._number_of_processed_lines
+
+    tags = [numpy.array("line-" + str(offset + i), dtype=self._tag_dtype, ndmin=1) for i in range(chunk_size)]
+    tags = numpy.concatenate(tags)
+
+    self.hdf_file["seqTags"][offset: offset + chunk_size] = tags
+
+  def _write_data_indices(self, chunk, side):
+    indices = numpy.concatenate(chunk)
+
+    if side == "source":
+      dataset = self.hdf_file["inputs/"]
+    else:
+      dataset = self.hdf_file["targets/data/classes"]
+
+    offset = self._write_offsets[side]
+    length = len(indices)
+    if offset + length > len(dataset):
+      buffer_size = max(self.data_buffer_size, length)
+      dataset.resize((offset + buffer_size,))
+
+    dataset[offset:offset + length] = indices
+
+    self._write_offsets[side] += length
+
+  def _finalize_data(self):
+    # Make sure the number of lines given by the user was correct.
+    # Otherwise lengths and labels would have trailing zeros.
+    assert self.number_of_lines == self._number_of_processed_lines, "Fewer lines ({}) in the corpus files " \
+        "than specified ({}).".format(self._number_of_processed_lines, self.number_of_lines)
+
+    # Trim datasets to actually occupied length, i.e. remove unused reserved space.
+    self.hdf_file["inputs"].resize((self._write_offsets["source"],))
+    self.hdf_file["targets/data/classes"].resize((self._write_offsets["target"],))
+
+  def _line_to_indices(self, line, side):
+    line = line.strip() + POSTFIX
+    words = line.split()
+
+    indices = [self._vocabularies[side].get(word, self._unknown_ids[side]) for word in words]
+
+    indices_numpy = numpy.array(indices, dtype=numpy.int32)
+    return indices_numpy
+
+  @staticmethod
+  def _open_file(file_name):
+    """
+    :param str file_name: filename of a plain text file, possibly zipped
+    :return: file handle
+    :rtype: TextIOWrapper|GzipFile
+    """
+    if file_name.endswith(".gz"):
+      open_function = gzip.open
+      mode = "rt"
```

Comment:
Why not directly `return gzip.open(file_name, "rt")`? And then `else:` and `return open(file_name, "r")`. Would be simpler and shorter and easier to read.